rfcs/0009-separate-wal.md

```diff
@@ -0,0 +1,143 @@
+
+- Support an optional separate object store dedicated specifically for WAL.
+- Provide a user-facing configuration mechanism.
+- Avoid significant refactorings.
+
+## Public API
+
+If user wants to use a separate WAL object store, the store instance should be
+provided while constructing a `Db`:
+
+```rust
+impl Db {
+    // ...
+
+    /// ...
+    ///
+    /// ## Arguments
+    /// ...
+    /// - `wal_object_store`: an optional [object store](ObjectStore) instance
+    ///   dedicated specifically for WAL.
+    ///
+    ///   If set to `None`, the passed `object_store` will be used for WAL storage.
+    ///
+    ///   NOTE: WAL durability and availability properties depend on the properties
+    ///         of the underlying object store. Make sure the configured object
+    ///         store is durable and available enough for your use case.
+    ///
+    /// ...
+
+    pub async fn open_with_opts<P: Into<Path>>(
+        path: P,
+        options: DbOptions,
+        object_store: Arc<dyn ObjectStore>,
+        wal_object_store: Option<Arc<dyn ObjectStore>>,
+    ) -> Result<Self, SlateDBError> {
+        // ...
+    }
+
+    // ...
+}
+```
+
+## Implementation
+
+The configured WAL object store is propagated down to `TableStore` instance
+which internally will select the appropriate object store for each operation
+based solely on the `SsTableId` involved in the operation. The object path
+structure will remain unchanged for both stores.
+
+There is already `transactional_wal_store` field in `TableStore`. The naming
+is misleading because apparently this store is used not only for WAL-related
+operations. The field should be renamed, the resulting field layout will look
+like this:
+
+```rust
+pub struct TableStore {
+    // ...
+
+    /// The main object store.
+    object_store: Arc<dyn ObjectStore>,
+    /// The dedicated WAL object store, if any.
+    wal_object_store: Option<Arc<dyn ObjectStore>>,
+
+    /// The transactional store wrapper of the main object store.
+    transactional_store: Arc<dyn TransactionalObjectStore>,
+    /// The transactional store wrapper of the dedicated WAL object store, if any.
+    wal_transactional_store: Option<Arc<dyn TransactionalObjectStore>>
+}
+```
+
+The object store selection procedure will look like this:
+
+```rust
+fn object_store_for(&self, id: &SsTableId) -> Arc<dyn ObjectStore> {
+    match id {
+        SsTableId::Wal(..) =>
+            if let Some(wal_object_store) = &self.wal_object_store {
+                wal_object_store.clone()
+            } else {
+                self.object_store.clone()
+            }
+        SsTableId::Compacted(..) => self.object_store.clone(),
+    }
+}
+```
+
+The transactional object store selection logic would look about the same.
+
+### Block and Object Cache
+
+WAL requires no caching support since it is read only once at startup. All
+operations on the dedicated WAL object store should be direct and bypass all
+the caches.
+
+### Cloning Support
```

Comment:
[non blocking suggestion] do we want to consider a defensive strategy to avoid accidentally opening a db instance that was previously operating with a separate WAL storage with only a main OS configured? 

basically I'm worried that you could conceivably have a roll back scenario that would revert to a slateDB deployment that used only a single OS after having already processed some data with an external WAL. Would be nice to fail hard in that situation so you don't corrupt the data.

One such implementation would be to just flag it in the manifest and throw an exception on startup if configured differently. Eventually this flag can be the full URI of the other object store or something like that.

Reply:
This was the scenario I had in mind as well when I wrote [this](https://github.com/slatedb/slatedb/pull/533#discussion_r2030028202) comment.
There is another [comment](https://github.com/slatedb/slatedb/pull/533#discussion_r2031541072) as well on this.

I do see that it is complex and out of scope for this RFC.
Shall we at least add a flag in the manifest that WAL is in a different store? When that flag is set, a different object store must be provided.

Reply:
WDYT about adding `wal_object_store_uri: string = null` field into `ManifestV1`? `null` would mean no separate WAL  store configured, empty string would stand for some store (the flag) which eventually would transition into a proper non-empty URI.

Reply:
I like this. I think it's a byte format compatible change, too.